app/controllers/declaration/TaricCodeRemoveController.scala

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2020 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.declaration
+
+import controllers.actions.{AuthAction, JourneyAction}
+import controllers.navigation.Navigator
+import forms.common.YesNoAnswer
+import forms.common.YesNoAnswer.{form, YesNoAnswers}
+import forms.declaration.{NactCode, TaricCode}
+import javax.inject.Inject
+import models.requests.JourneyRequest
+import models.{DeclarationType, ExportsDeclaration, Mode}
+import play.api.data.Form
+import play.api.i18n.I18nSupport
+import play.api.mvc._
+import services.cache.ExportsCacheService
+import uk.gov.hmrc.play.bootstrap.controller.FrontendController
+import views.html.declaration.taric_code_remove
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class TaricCodeRemoveController @Inject()(
+  authenticate: AuthAction,
+  journeyType: JourneyAction,
+  override val exportsCacheService: ExportsCacheService,
+  navigator: Navigator,
+  mcc: MessagesControllerComponents,
+  taricCodeRemove: taric_code_remove
+)(implicit ec: ExecutionContext)
+    extends FrontendController(mcc) with I18nSupport with ModelCacheable {
+
+  def displayPage(mode: Mode, itemId: String, code: String): Action[AnyContent] = (authenticate andThen journeyType) { implicit request =>
+    Ok(taricCodeRemove(mode, itemId, code, YesNoAnswer.form()))
+  }
+
+  def submitForm(mode: Mode, itemId: String, code: String): Action[AnyContent] = (authenticate andThen journeyType).async { implicit request =>
+    form()
+      .bindFromRequest()
+      .fold(
+        (formWithErrors: Form[YesNoAnswer]) => Future.successful(BadRequest(taricCodeRemove(mode, itemId, code, formWithErrors))),
+        formData => {
+          formData.answer match {
+            case YesNoAnswers.yes =>
+              updateExportsCache(itemId, request.cacheModel.itemBy(itemId).flatMap(_.taricCodes).getOrElse(Seq.empty).filterNot(_.taricCode == code))
```

Comment:
I think it can be worth to move this logic into updateExportsCache method.

Pass there code and do everything inside (not sure if it's the best idea, because of the method responsibility).

If it's not good idea at least we can extract preparing the data into variable to make it more readable

Reply:
I don't get you scala bods sometimes.  I thought doing it all in a single statement was cool and readability was for wimps.

Reply:
Hmm, I think everything is based on the situation, if you will need to pass just parties like `request.cacheModel.parties` that's okay, it's short and probably readable enought to be in the same line.

`request.cacheModel.itemBy(itemId).flatMap(_.taricCodes).getOrElse(Seq.empty).filterNot(_.taricCode == code)` This is quite big and manipulate the data, this is the reason that I thought it can be useful to extract this as a variable.

I approved PR, so this was just a non blocking proposal :)

Reply:
Well I'm going to change it.  And same code in NACT controller.  So you'll have to review all over again.  :-)

Reply:
Done